### PR TITLE
Improve inlining of BitBlock fns and iterators

### PR DESCRIPTION
**Pull request by cuviper**
The BitBlock fns can be mostly optimized away, but only if they're inlined.
Running the benchmarks from an extern crate shows the difference.
Adding explicit `#[inline]` attributes lets them inline across crate boundaries too.

### CODE REVIEW

**Comment by cuviper** [comment]
Here are some specific numbers before inlining:
```
$ cargo bench --features=nightly bench::
   Compiling bit-vec v0.4.1 (file:///home/jistone/rust/bit-vec)
     Running target/release/bit_vec-514c687ce02b23dd

running 8 tests
test bench::bench_bit_set_big_fixed    ... bench:       1,908 ns/iter (+/- 33)
test bench::bench_bit_set_big_variable ... bench:       3,943 ns/iter (+/- 1,584)
test bench::bench_bit_set_small        ... bench:       1,868 ns/iter (+/- 151)
test bench::bench_bit_vec_big_iter     ... bench:      11,975 ns/iter (+/- 324)
test bench::bench_bit_vec_big_union    ... bench:         472 ns/iter (+/- 11)
test bench::bench_bit_vec_small_iter   ... bench:         199 ns/iter (+/- 7)
test bench::bench_from_elem            ... bench:         379 ns/iter (+/- 9) = 5403 MB/s
test bench::bench_usize_small          ... bench:       1,783 ns/iter (+/- 49)

test result: ok. 0 passed; 0 failed; 0 ignored; 8 measured

     Running target/release/extern-4139b10cfb20f415

running 8 tests
test bench::bench_bit_set_big_fixed    ... bench:       3,501 ns/iter (+/- 147)
test bench::bench_bit_set_big_variable ... bench:       5,498 ns/iter (+/- 2,292)
test bench::bench_bit_set_small        ... bench:       3,566 ns/iter (+/- 162)
test bench::bench_bit_vec_big_iter     ... bench:     359,797 ns/iter (+/- 11,210)
test bench::bench_bit_vec_big_union    ... bench:         459 ns/iter (+/- 9)
test bench::bench_bit_vec_small_iter   ... bench:       7,138 ns/iter (+/- 329)
test bench::bench_from_elem            ... bench:         380 ns/iter (+/- 33) = 5389 MB/s
test bench::bench_usize_small          ... bench:       1,780 ns/iter (+/- 26)

test result: ok. 0 passed; 0 failed; 0 ignored; 8 measured
```

With inlining:
```
$ cargo bench --features=nightly bench::
   Compiling bit-vec v0.4.1 (file:///home/jistone/rust/bit-vec)
     Running target/release/bit_vec-514c687ce02b23dd

running 8 tests
test bench::bench_bit_set_big_fixed    ... bench:       1,909 ns/iter (+/- 851)
test bench::bench_bit_set_big_variable ... bench:       3,943 ns/iter (+/- 120)
test bench::bench_bit_set_small        ... bench:       1,860 ns/iter (+/- 49)
test bench::bench_bit_vec_big_iter     ... bench:      11,939 ns/iter (+/- 348)
test bench::bench_bit_vec_big_union    ... bench:         471 ns/iter (+/- 20)
test bench::bench_bit_vec_small_iter   ... bench:         199 ns/iter (+/- 9)
test bench::bench_from_elem            ... bench:         385 ns/iter (+/- 15) = 5319 MB/s
test bench::bench_usize_small          ... bench:       1,777 ns/iter (+/- 65)

test result: ok. 0 passed; 0 failed; 0 ignored; 8 measured

     Running target/release/extern-4139b10cfb20f415

running 8 tests
test bench::bench_bit_set_big_fixed    ... bench:       1,916 ns/iter (+/- 100)
test bench::bench_bit_set_big_variable ... bench:       3,956 ns/iter (+/- 75)
test bench::bench_bit_set_small        ... bench:       1,888 ns/iter (+/- 61)
test bench::bench_bit_vec_big_iter     ... bench:      19,554 ns/iter (+/- 507)
test bench::bench_bit_vec_big_union    ... bench:         471 ns/iter (+/- 13)
test bench::bench_bit_vec_small_iter   ... bench:         389 ns/iter (+/- 13)
test bench::bench_from_elem            ... bench:         381 ns/iter (+/- 21) = 5375 MB/s
test bench::bench_usize_small          ... bench:       1,825 ns/iter (+/- 84)

test result: ok. 0 passed; 0 failed; 0 ignored; 8 measured
```

So external use is still a little slower, but it's much closer.

**Comment by cuviper** [comment]
One more thing, I think RNG is filling a big chunk of the benchmarks.  A simple `time cargo bench ...` shows user vs. sys time is about half and half, and `strace -c` shows these are mostly `getrandom()` syscalls.

What would you think about switching to `weak_rng()` for the benchmarks?

**Comment by pczarn** [comment]
@cuviper: Yes, `weak_rng()` is the right tool. Ideally, calls to `r.next_u32` would be outside the loops.

`release` doesn't perform LTO. Could it be that external use is a little slower, because `fn from_fn`, or something else is not inlined?

**Comment by cuviper** [comment]
@pczarn: Only `bench_bit_vec_big_iter` and `bench_bit_vec_small_iter` are still substantially different, and I don't see any extra calls in their disassembled functions to make me suspect missed inlining.  I'll do some profiling and see where they spend time.

**Comment by cuviper** [comment]
OK, the performance difference in iterators appears to be from the `TRUE` and `FALSE` globals referenced by the `BitVec::index()`.  For the internal benchmark, those get optimized as constants, but in the external benchmark they are used as variables.  The optimizer does load those values into registers outside of the loop, but still the code can't be as compact as the constant case.

I just tried changing `Iter::next()` to use `get()` instead, as `index()` does internally, but strangely this reduced performance!  It was nicely equal performance between internal and external benchmarks, but equally slower is not nice.  I'll keep trying...

**Comment by cuviper** [comment]
The current `next()` is `self.range.next().map(|i| self.bit_vec[i])`
It's slower as `.and_then(|i| self.bit_vec.get(i))`
But it's good even externally with `.map(|i| self.bit_vec.get(i).unwrap())`

I'll push additional commits and post new benchmarks shortly.

**Comment by cuviper** [comment]
```
$ cargo bench --features nightly bench::
     Running target/release/bit_vec-514c687ce02b23dd

running 8 tests
test bench::bench_bit_set_big_fixed    ... bench:         311 ns/iter (+/- 8)
test bench::bench_bit_set_big_variable ... bench:         723 ns/iter (+/- 17)
test bench::bench_bit_set_small        ... bench:         301 ns/iter (+/- 9)
test bench::bench_bit_vec_big_iter     ... bench:      11,944 ns/iter (+/- 441)
test bench::bench_bit_vec_big_union    ... bench:         472 ns/iter (+/- 20)
test bench::bench_bit_vec_small_iter   ... bench:         191 ns/iter (+/- 7)
test bench::bench_from_elem            ... bench:         408 ns/iter (+/- 22) = 5019 MB/s
test bench::bench_usize_small          ... bench:         194 ns/iter (+/- 9)

test result: ok. 0 passed; 0 failed; 0 ignored; 8 measured

     Running target/release/extern-4139b10cfb20f415

running 8 tests
test bench::bench_bit_set_big_fixed    ... bench:         325 ns/iter (+/- 14)
test bench::bench_bit_set_big_variable ... bench:         742 ns/iter (+/- 22)
test bench::bench_bit_set_small        ... bench:         294 ns/iter (+/- 9)
test bench::bench_bit_vec_big_iter     ... bench:      11,946 ns/iter (+/- 266)
test bench::bench_bit_vec_big_union    ... bench:         473 ns/iter (+/- 20)
test bench::bench_bit_vec_small_iter   ... bench:         272 ns/iter (+/- 15)
test bench::bench_from_elem            ... bench:         390 ns/iter (+/- 3) = 5251 MB/s
test bench::bench_usize_small          ... bench:         181 ns/iter (+/- 12)

test result: ok. 0 passed; 0 failed; 0 ignored; 8 measured
```

File: src/lib.rs [hunk 1]
```diff
@@ -2072,113 +2080,5 @@ mod tests {
...
-        let bit = black_box(true);
-        b.iter(|| {
-            // create a BitVec and popcount it
-            BitVec::from_elem(cap, bit).blocks()
-                                       .fold(0, |acc, b| acc + b.count_ones())
-        });
-        b.bytes = cap as u64 / 8;
-    }
-}
+#[cfg(all(test, feature = "nightly"))] mod bench;
```
Comment (pczarn) [inline src/lib.rs]: nit: this should be in two lines.

**Comment by pczarn** [comment]
Could you move `bench.rs` to `benches`?

**Comment by cuviper** [comment]
I believe cargo will build each file in `benches/` as a separate benchmark program, so I'm not sure how to move the benchmarks there and still have them usable as in-crate benchmarks.

File: src/lib.rs (hunk 1 above)
Comment (cuviper) [inline src/lib.rs]: OK, I can split it.  I only wrote it that way to match the one-liners at the top of `src/lib.rs`.

**Comment by pczarn** [comment]
Looks good, thanks!

File: src/lib.rs (hunk 1 above)
Comment (pczarn) [inline src/lib.rs]: Nevermind.

**Pull request merged**